File: thinking-categorically/applicative-functor/applicative-functor-laws/interchange-law.ts
function main() {
  // List of functions
  const u: Array<(x: number) => number> = [
    x => x + 1,
    x => x * 2
  ];

  const y = 3;

  // Left side: u <*> pure y
  const left = u.map(f => f(y));

  // Right side: pure (f => f(y)) <*> u
  const right = [(f: (x: number) => number) => f(y)].flatMap(g => u.map(f => g(f)));

  console.log("Left:  ", left);   // [4, 6]
  console.log("Right: ", right);  // [4, 6]
  console.log("Equal? ", arraysEqual(left, right)); // true
}

function arraysEqual<T>(a: T[], b: T[]): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}

main();

File: thinking-categorically/applicative-functor/applicative-functor-laws/interchange-law.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main()
    {
        // List of functions (like [(+1), (*2)])
        var u = new List<Func<int, int>> { x => x + 1, x => x * 2 };

        // A pure value
        int y = 3;

        // Left side: u <*> pure y
        var left = u.Select(f => f(y)).ToList();

        // Right side: pure (f => f y) <*> u
        var right = new List<Func<Func<int, int>, int>> { f => f(y) }
            .SelectMany(g => u.Select(f => g(f)))
            .ToList();

        Console.WriteLine($"Left:  {string.Join(", ", left)}");
        Console.WriteLine($"Right: {string.Join(", ", right)}");
        Console.WriteLine($"Equal? {left.SequenceEqual(right)}");
    }
}
